src/efivarfs.c

```diff
@@ -28,18 +29,36 @@
 #endif
 
 static char const default_efivarfs_path[] = "/sys/firmware/efi/efivars/";
+static char *efivarfs_path;
 
 static char const *
 get_efivarfs_path(void)
```

Comment:
It looks like this purpose of this function was to support lazy initialization, but if you're going to use a CONSTRUCTOR to initialize `efivarfs_path`, then this function seems unnecessary.  It's contents could be put into the CONSTRUCTOR function instead.  This removes need to check whether `efivarfs_path` is NULL both in the constructor and any time it is being used.

Reply:
I'm not sure it does actually obviate that—though admittedly the situation where it doesn't is fraught with bad ideas.  Consider a program which links against `libefivar.so` and another library, `libmistakes.so`.  `libmistakes.so` does not link against `libefivar.so` or have any `DT_NEEDED` object references to it, but has a `DT_INIT` constructor that does something like:
```
static void CONSTRUCTOR init_mistakes(void)
{
  int (*gnvn)(efi_guid_t **guid, char **name);
  char *name;
  efi_guid_t *guid;
  int rc;

  gnvn = dlsym(RTLD_DEFAULT, "efi_get_next_variable_name");
  if (!gnvn)
    return;
  rc = gnvn(&guid, &name);
  if (rc >= 0)
    printf("first EFI variable is %s\n", name);
}
```
Since `libmistakes.so` has no `DT_NEEDED` pointing at `libefivar.so`, the `DT_INIT` sorting is not required to ensure that `init_mistakes()` is called after `init_efivarfs_path()`; they could be called in any order.  Though that does mean the error check on the allocation also needs to be in `get_efivarfs_path()`, rather than `init_efivarfs_path()`.  I've fixed that in 3d19e7d492a.

I realize this sounds completely contrived but the basic outline here is more or less how `kdeinit` works...

Anyway, that aside, are you okay with what I've done in https://github.com/rhboot/efivar/pull/155/commits/77f77a4c6d6 ?